### ISSUE TITLE
[a11y] Do not exclusively use color to convey information

### ISSUE THREAD

**Issue by manuhabitela** [gristlabs/grist-core#1244]
### Describe the problem to be solved

👋

The goal of this issue is to track all issues concerning the exclusive usage of color to convey information. This is something we should not do, according the WCAG, the international accessibility guidelines.

Context is:

Color only should not be used to convey info. For example, an "online status" badge being shown with a green background when online, and a red background when offline, is not enough: If I have trouble distinguishing colors, it'll hard for me to understand if I'm online or not. To fix this example, I could change the text to "online" or "offline", or add an icon that changes (with a crossed wifi icon when offline).

Sources:

https://www.w3.org/WAI/WCAG21/Understanding/use-of-color

### Describe the solution you would like

Here are the issues I noticed for now that we can discuss:

- in the documents list, switching between list view and gallery view can be misleading. Current viewing mode is shown only with a color change of the buttons. Since there are only two buttons, it can be tricky to understand which is enabled. _And_ the current [color issues](https://github.com/gristlabs/grist-core/issues/1243) makes it so that the two colors are not contrasted enough. I can have trouble noticing a color change. Here, adding a background and/or a border would fix the issue.
- we can argue toggle buttons can be difficult to understand. Which visual state means "enabled" and which means "disabled"? We could imagine adding a checkbox icon when the button is enabled.
- same feedback could be made on grouped buttons (that act as radio inputs). When choosing between 3 choices and more the issue is not as pronounced; but when choosing between 2 choices, the question remains: which is enabled, which is disabled?
- the Table/Column tabs are misleading from my point of view. The inactive tab has bold text, while the active tab has an accent background color.

Screenshots:

- ![image](https://github.com/user-attachments/assets/76eb6813-216b-49b7-a595-5272e1667252)
- ![image](https://github.com/user-attachments/assets/5edc9b6a-7bd7-461b-92fd-e48e6ba14814)
- ![image](https://github.com/user-attachments/assets/19225fc6-3176-4936-aab8-7b3d75d014cc)
- ![image](https://github.com/user-attachments/assets/1cf40596-0785-4edc-a0d4-3d6789a17c4f)
- ![image](https://github.com/user-attachments/assets/9aff24ea-f4e0-4d70-bb51-b2aa2afb18dd)
- ![image](https://github.com/user-attachments/assets/dfb291d2-7f4b-4450-baa4-1a19e93578ef)

**Comment by tayflo** [comment]
Hello there :wave: If I might share my humble opinion, as an enthusiastic Grist user:

1. I perfectly agree that List/Gallery views buttons, in document list, right now have an enabled/disabled state which is only conveyed through colors which are not distinctive/contrasted enough from each other.

2. I also agree that the Table/Column tabs might feel misleading when starting using Grist (especially the bold on the inactive tab). Your proposition, resulting in this [approved proposal](https://github.com/gristlabs/grist-core/issues/1011#issuecomment-2415561113) is indeed clearer in my opinion.

3. I'm not so sure about the toggle buttons tho. When an information is based on background-to-foreground-switching (with sufficient contrast between background and foreground, e.g. dark-on-light to light-on-dark), it is not "only conveyed through color", as there are sufficient contrast to make the distinction between the two states (it's the same for point 2 tho, but I still find your proposal clearer).

This last point is mentioned in the [WCAG guideline](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color) you quoted itself:

> If content is conveyed through the use of colors that differ not only in their hue, but that also have a significant difference in lightness, then this counts as an additional visual distinction (...).
> Similarly, if content is distinguished by inverting an element's foreground and background colors, this would pass (again, assuming that the foreground and background colors have a sufficient contrast).

So to me the problem here is not really about "conveyed only by color", but more, as you state it, how to know which correspond to "enabled" and which correspond to "disabled" (and it affects users independently of potential visual deficiencies or colorblindess).

It is arguable but generally speaking I don't find it *too* confusing that light-on-dark means "enabled" whereas dark-on-light means "disabled", in an interface where things are generally dark-on-light.

It might maybe be improved tho. Below are some ideas about how to achieve that. I will take examples from the [Material Design](https://m3.material.io/) (a rulebook and resource pack for software design which is developped by Google through [open-source repo](https://www.github.com/material-components), is used by many, and [champions accessibility](https://m3.material.io/foundations/overview/principles) as a core principle).

1. The solution you mention: add a supplementary icon on the button. I'm not sure I'm a big fan of this option, I feel like it might clutter a bit the interface.

   Nonetheless, it is an option considered in Material Design for [selected elements](https://m3.material.io/foundations/interaction/selection#7d4bdaf6-2d28-45fd-9225-9235457b7e05) (for [enabled elements](
https://m3.material.io/foundations/interaction/states/applying-states#39b2fc90-01db-41b5-b6f8-47be61ed1479), a light-on-dark pattern is used, but I think it might be "enabled" in opposition to inactive/unclickable elements); here is how it looks like:

   ![m1qhf1tb-1](https://github.com/user-attachments/assets/e3173178-60b7-4d61-b9f4-ab15ca9682bb)

2. Another solution would be to add an inner shadow to elements. But since the elements are already very dark, I don't feel like it is achievable here. Or both "enabled" and "disabled" states should be light-on-dark, with "enabled" state with an inner shadow.

3. The last solution I think of (which would have my preference I think) would be to limit the dark-background to a subset of the selector; that is, keep the original border and keep a light margin (same tone for each elements in the selector) around the dark background. That way, the selected element is clearly a subset of the selector, whereas non-selected elements looks within the broarder frame of the selector.

   Here is how it looks like in Google Docs:

   ![2024-10-20T18-19](https://github.com/user-attachments/assets/1a49d688-6a8f-40e7-be0a-600ae8ac2dfa)

   And a very not-nice representation of mine of the idea on the TextBox/Spinner selector :-)

   ![2024-10-20T19-06](https://github.com/user-attachments/assets/3224ef0d-a8ce-4193-80e0-9a2e7138c06e)

## Additionnal notes about properly handling color accessibility

- I was wondering if colorblindness was mainly about [hue](https://en.wikipedia.org/wiki/HSL_and_HSV) rather than lightness and saturation, so I double-checked. It seems to be the case, but lightness and saturation can also enter in play (that's why in the end, what matters is contrast between colors). For instance ligh-yellow or light-turquoise could be hardly distinguishable from white, and dark-red or dark-blue from black. (Sources: tools listed below, [Wikipedia](https://en.wikipedia.org/wiki/Luminous_efficiency_function#Color_blindness), testimonials on [Reddit](https://www.reddit.com/r/ColorBlind/), Quora...)

- Some tools and techniques can be used. [Wikipedia](https://en.wikipedia.org/wiki/Color_blindness#Digital_design) lists some usual techniques.

- There are some ways to help building color palettes which are adapted to colorblind people:

  - Some [pre-defined palettes](https://en.wikipedia.org/wiki/Color_blindness#Ordered_Information) exist (e.g. Viridis).

  - The HCT color space [from Material Design](https://m3.material.io/styles/color/system/how-the-system-works) is thought to help build constrated-enough color palettes.

  - This [little tool](https://gka.github.io/palettes/) also checks for "color-blind safety" when generating gradient.

  - The [tailwind.ink](https://tailwind.ink/) tool may also be interesting.

- There are some ways to help visualising and check designs against that:

  - A quick (but imperfect) way to check for contrast is to switch on grayscale on one's monitor or operating system (imperfect because not only hue is implied, see the first point I mentioned in this section).

  - There are tools to visualize how an image might look for colorblind people, such as on [daltonlens.org](https://daltonlens.org/colorblindness-simulator) and [myndex.com](https://www.myndex.com/CVD/).

  - There are tools built-in in design softwares and browsers. For instance you can turn this on on [Firefox](https://firefox-source-docs.mozilla.org/devtools-user/accessibility_inspector/simulation/index.html).

- There are also tools to check for contrast between colors; already listed in issue #1243.

**Comment by manuhabitela** [comment]
Hey, thanks for all your input, it's greatly appreciated :)

Yeah the whole color contrast stuff is sometimes not that straightforward. Current WCAG calculations also have their limits. [This article](https://blog.datawrapper.de/color-contrast-check-data-vis-wcag-apca/) is interesting if you are curious.

**Comment by lusebille** [comment]
Link to mocks https://www.figma.com/design/wcpetFt6aOKzTszcvPPWLQ/%5B05%2F24%5D-Grist-Design?node-id=2438-131024&t=qX4iHFYd8uzsW8kZ-1

**Comment by dsagal** [comment]
This is an umbrella issue that will include many fixes/improvements. These ones, present in the linked mocks, are ready to work on:
- Styling of creator-panel tabs (for switching between Table and Column configuration)
- Styling of selectors in the creator panel (e.g. TextBox vs Spinner, alignment) and similarly-styled sort toggle in the app home.

**Comment by manuhabitela** [comment]
Actually @dsagal your comment on the figma is a good feedback on something that I missed (sorry @lusebille, completely my mistake :see_no_evil:).

For solo toggle icons like the text wrapping button or linked label/status icon button, our new proposal is not perfect:

![image](https://github.com/user-attachments/assets/bcd3b002-2294-4374-ae8c-f4812b3762d1)

Because the contrast between the active border and inactive border is not enough.

This is not an issue on group buttons and toggles like the "Cell format" or "Text box / Spinner" inputs, because the selected item has borders that others items do not, and the borders color contrast ratio is enough with the white background.

Ideas to fix:

 - we could keep these visuals and fix this issue by having a different icon when a toggle is on or off
 - scratch that idea of the light background when a toggle is on, and go back to an idea where a toggled on item is more visually marked (like the previous black color)
 - have a design more looking like the one where there are multiple toggles, with an inner border, I think Lucie that was one thing you suggested at the beginning? and I was the one suggesting another, less accessible version… oops:
![image](https://github.com/user-attachments/assets/a2be8d9b-8197-44fe-a658-e6da6b3de552)

**Comment by lusebille** [comment]
I suggest changing icon for the 'link' icon , and for the one below( the 'go to the line' button ) using the extra border , it makes more sense for this one because it's next to the text algin option ( adding the extra border to the link/unlink icon will be too much, just changing icon should be enough, and it's the common behavior in tools ) >> I've made the change on the mock
